### PR TITLE
[containerapp] Support keda msi

### PR DESCRIPTION
**Pull request by njuCZ**
---

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

- [ ] Have you run `azdev style <YOUR_EXT>` locally? (`pip install azdev` required)
- [ ] Have you run `python scripts/ci/test_index.py -q` locally? (`pip install wheel==0.30.0` required)
- [ ] My extension version conforms to the [Extension version schema](https://github.com/Azure/azure-cli/blob/release/doc/extensions/versioning_guidelines.md)

For new extensions:

- [ ] My extension description/summary conforms to the [Extension Summary Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/extensions/extension_summary_guidelines.md).

### About Extension Publish

There is a pipeline to automatically build, upload and publish extension wheels.  
Once your pull request is merged into main branch, a new pull request will be created to update `src/index.json` automatically.  
You only need to update the version information in file setup.py and historical information in file HISTORY.rst in your PR but do not modify `src/index.json`.

### CODE REVIEW

**Comment by yonzhan** [comment]
containerapp

File: src/containerapp/azext_containerapp/containerapp_decorator.py [hunk 1]
```diff
@@ -975,6 +976,17 @@ def set_up_runtime(self):
                         "java": runtime_java_def
                     }
             safe_set(self.containerapp_def, "properties", "configuration", "runtime", value=runtime_def)
+    
+    # pylint: disable=unsupported-assignment-operation
+    def set_up_scale_rule(self):
```
Comment (Greedygre) [inline src/containerapp/azext_containerapp/containerapp_decorator.py:981]: Should we add validation logic about the `--scale-rule-identity` is only for a custom scale rule?

File: src/containerapp/azext_containerapp/containerapp_decorator.py [hunk 2]
```diff
@@ -975,6 +980,17 @@ def set_up_runtime(self):
...
+        scale_def = super().set_up_scale_rule()
+        if scale_def and scale_def["rules"] and scale_def["rules"][0] and scale_def["rules"][0]["custom"] and self.get_argument_scale_rule_identity():
+            identity = self.get_argument_scale_rule_identity().lower()
+            if identity != "system":
+                subscription_id = get_subscription_id(self.cmd.cli_ctx)
```
Comment (Greedygre) [inline src/containerapp/azext_containerapp/containerapp_decorator.py:990]: What if the identity resourceId not in current subscription?
The `_ensure_identity_resource_id` change the subscription Id to current sub, which is not expect.
image

**Review by Greedygre — APPROVED** [review]

File: src/containerapp/azext_containerapp/_params.py [hunk 3]
```diff
@@ -306,6 +306,11 @@ def load_arguments(self, _):
         c.argument('managed_env', validator=validate_env_name_or_id, options_list=['--environment'], help="Name or resource ID of the container app's environment.")
         c.argument('environment_type', arg_type=get_enum_type(["managed", "connected"]), help="Type of environment.", is_preview=True)
 
+    # scale
+    with self.argument_context('containerapp', arg_group='Scale') as c:
```
Comment (Juliehzl) [inline src/containerapp/azext_containerapp/_params.py:310]: could you move to L298. Add this parameter with other scale parameters

File: src/containerapp/azext_containerapp/containerapp_decorator.py [hunk 4]
```diff
@@ -646,6 +647,10 @@ def validate_arguments(self):
         if self.get_argument_service_bindings() and len(self.get_argument_service_bindings()) > 1 and self.get_argument_customized_keys():
             raise InvalidArgumentValueError("--bind have multiple values, but --customized-keys only can be set when --bind is single.")
         validate_runtime(self.get_argument_runtime(), self.get_argument_enable_java_metrics(), self.get_argument_enable_java_agent())
+        if self.get_argument_scale_rule_type() and self.get_argument_scale_rule_identity():
```
Comment (Juliehzl) [inline src/containerapp/azext_containerapp/containerapp_decorator.py:650]: do you have similar validation for update scenario? For example, customer already has scale rule and use update command to specify scale rule identity, will it fail?

File: src/containerapp/azext_containerapp/_params.py (hunk 3 above)
Comment (njuCZ) [reply src/containerapp/azext_containerapp/_params.py:310]: updated

File: src/containerapp/azext_containerapp/containerapp_decorator.py (hunk 4 above)
Comment (njuCZ) [reply src/containerapp/azext_containerapp/containerapp_decorator.py:650]: there is similar validation for update scenario

File: src/containerapp/azext_containerapp/containerapp_decorator.py [hunk 5]
```diff
@@ -1095,6 +1121,15 @@ def construct_payload(self):
             safe_set(self.new_containerapp, "properties", "configuration", "maxInactiveRevisions", value=self.get_argument_max_inactive_revisions())
         self.set_up_runtime()
 
+        if self.get_argument_scale_rule_name() and self.get_argument_scale_rule_identity():
+            customScaleRule = safe_get(self.new_containerapp, "properties", "template", "scale", "rules", default=[])
```
Comment (Greedygre) [inline src/containerapp/azext_containerapp/containerapp_decorator.py:1125]: customScaleRule -> custom_scale_rule

**Review by Juliehzl — APPROVED** [review]

File: src/containerapp/azext_containerapp/containerapp_decorator.py [hunk 6]
```diff
@@ -1095,6 +1121,15 @@ def construct_payload(self):
...
+            if customScaleRule and customScaleRule[0] and customScaleRule[0]["custom"]:
```
Comment (Greedygre) [inline src/containerapp/azext_containerapp/containerapp_decorator.py:1126]:
```suggestion
           custom_scale_rule = safe_get(self.new_containerapp, "properties", "template", "scale", "rules", default=[])
            if custom_scale_rule and len(custom_scale_rule) > 0 and custom_scale_rule[0]["custom"]:
```

**Review by Greedygre — APPROVED** [review]

File: src/containerapp/azext_containerapp/_help.py [hunk 7]
```diff
@@ -892,6 +892,18 @@
...
+    - name: Create a container app with a custom scale rule using identity to authenticate
+      text: |
+          az containerapp create -n my-containerapp -g MyResourceGroup \\
+              --image my-queue-processor --environment MyContainerappEnv \\
+              --user-assigned myUserIdentityResourceId --min-replicas 4 --max-replicas 8 \\
+              --scale-rule-name queue-based-autoscaling \\
+              --scale-rule-type azure-queue \\
+              --scale-rule-metadata "accountName=mystorageaccountname" \\
+                                    "cloud=AzurePublicCloud" \\
+                                    "queueLength": "5" "queueName": "foo" \\
```
Comment (yanzhudd) [inline src/containerapp/azext_containerapp/_help.py:904]: please check if the example of this param is correct or not.

Comment (njuCZ) [reply src/containerapp/azext_containerapp/_help.py]: updated

**Review by Greedygre — APPROVED** [review]

**Review by yanzhudd — APPROVED** [review]

**Pull request merged**